File: Reqnroll/Chapter15/15.4-TestActionPattern/After/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(AuthenticationContext authContext, OrderingApiDriver orderingApiDriver, NotificationsApiDriver notificationsApiDriver)
{
    private int? placedOrderNo;

    [Given("they have placed an order")]
    public async Task GivenTheyHavePlacedAnOrder()
    {
        var placeOrderRequest = new PlaceOrderRequestObjectMother().Build();
        var placedOrder = await orderingApiDriver.PlaceOrder(placeOrderRequest).Execute();
        placedOrderNo = placedOrder.OrderNo;
    }

    [When("they cancel the placed order")]
    public async Task WhenTheyCancelThePlacedOrder()
    {
        int orderNumber = placedOrderNo ?? throw new InvalidOperationException("No placed order");
        await orderingApiDriver.CancelOrder(orderNumber).Execute();
    }

    [Then("they should receive a notification about the cancellation")]
    public async Task ThenTheyShouldReceiveANotificationAboutTheCancellation()
    {
        string customerName = authContext.LoggedInCustomerName ?? throw new InvalidOperationException("No logged in customer name");
        var notifications = await notificationsApiDriver.GetNotifications(customerName).Execute();

        Assert.IsNotNull(notifications);
        Assert.IsTrue(notifications.Any(n =>
                n.Message.Contains("cancelled", StringComparison.OrdinalIgnoreCase)),
            "Expected a cancellation notification but none was found.");
    }
}

File: CucumberJVM/Chapter15/15.3-InfrastructureLayerPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

import java.util.Optional;

public class OrderingStepDefinitions {
    private final OrderingApiDriver orderingApiDriver;

    private Integer placedOrderNo;

    public OrderingStepDefinitions(OrderingApiDriver orderingApiDriver) {
        this.orderingApiDriver = orderingApiDriver;
    }

    @Given("they have placed an order")
    public void theyHavePlacedAnOrder() throws Exception {
        var placeOrderRequest = new PlaceOrderRequestObjectMother().build();
        var placedOrder = orderingApiDriver.placeOrder(placeOrderRequest);
        placedOrderNo = placedOrder.getOrderNo();
    }

    @When("they cancel the placed order")
    public void theyCancelTheOrder() throws Exception {
        int orderNo = Optional.ofNullable(placedOrderNo).orElseThrow(() -> new RuntimeException("No placed order"));
        orderingApiDriver.cancelOrder(orderNo);
    }
}
